File: BOJ/11505번/11505.cpp
#include <iostream>
#include <vector>
#include <cmath>
using namespace std;
using lint = long long;
lint mod = 1000000007;

vector<lint> tree;
lint cal(lint a, lint b)
{
    return a * b % mod;
}

void init(int size)
{
    for (int i = size - 1; i > 0; i--)
    {
        tree[i] = cal(tree[i * 2], tree[i * 2 + 1]);
    }
}

lint sum(int node, int left, int right, int start, int end)
{
    if (left > end || right < start) return 1;
    if (left <= start && end <= right) return tree[node];
    int mid = (start + end) / 2;
    return cal(sum(node * 2, left, right, start, mid), sum(node * 2 + 1, left, right, mid + 1, end));
}

void update(int index, int size, lint num)
{
    index += size;
    tree[index] = num;
    while (index > 1)
    {
        index /= 2;
        tree[index] = cal(tree[index * 2], tree[index * 2 + 1]);
    }
}

int main(void)
{
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);
    int N, M, K;
    cin >> N >> M >> K;

    int height = (int) ceil(log2(N));
    int tree_size = (1 << (height + 1));
    int leaf_size = (1 << height);
    tree.assign(tree_size, 1);

    for (int i = 0; i < N; i++) cin >> tree[leaf_size + i];
    init(leaf_size);
    
    M += K;
    lint A, B, C;
    while (M--)
    {
        cin >> A >> B >> C;
        if (A == 1) update(B - 1, leaf_size, C);
        else cout << sum(1, B - 1, C - 1, 0, leaf_size - 1) % mod << '\n';
    }
    return 0;
}

File: BOJ/11505번/11505.kt
import java.util.*
class SegmentTree(size: Int, array: LongArray) {
    private var treeSize = 1
    private var tree: LongArray
    init {
        while (treeSize < size) treeSize *= 2
        tree = LongArray(treeSize * 2) { 1 }
        for (it in 0 until size) tree[treeSize + it] = array[it]
        for (it in treeSize - 1 downTo 1) tree[it] = (tree[it * 2] * tree[it * 2 + 1]) % mod
    }
    fun update(index: Int, num: Long) {
        var it = index + treeSize
        tree[it] = num
        while (it > 1) {
            it /= 2
            tree[it] = (tree[it * 2] * tree[it * 2 + 1]) % mod
        }
    }
    fun getSum(left: Int, right: Int) = getSum(1, left, right, 0, treeSize - 1)
    private fun getSum(node: Int, left: Int, right: Int, start: Int, end: Int): Long {
        return when {
            left > end || right < start -> 1
            left <= start && end <= right -> tree[node]
            else -> {
                val mid = (start + end) / 2
                (getSum(node * 2, left, right, start, mid) * getSum(node * 2 + 1, left, right, mid + 1, end)) % mod
            }
        }
    }
    companion object {
        private const val mod = (1e9 + 7).toLong()
    }
}
fun main() = with(System.`in`.bufferedReader()) {
    val bw = System.out.bufferedWriter()
    var token = StringTokenizer(readLine())
    val n = token.nextToken().toInt()
    val m = token.nextToken().toInt() + token.nextToken().toInt()
    val array = LongArray(n)
    for (it in 0 until n) array[it] = readLine().toLong()
    val segmentTree = SegmentTree(n, array)
    for (it in 0 until m) {
        token = StringTokenizer(readLine())
        val a = token.nextToken().toInt()
        val b = token.nextToken().toInt()
        val c = token.nextToken().toLong()
        if (a == 1) segmentTree.update(b - 1, c)
        else bw.write("${segmentTree.getSum(b - 1, c.toInt() - 1)}\n")
    }
    bw.close()
}